packages/flutter/lib/src/material/combo_box.dart

```diff
@@ -0,0 +1,747 @@
+        : effectiveStyle;
+
+      final MenuItemButton menuItemButton = MenuItemButton(
+        style: effectiveStyle,
+        leadingIcon: entry.leadingIcon,
+        trailingIcon: entry.trailingIcon,
+        onPressed: entry.enabled ? () {
+          textEditingController.text = entry.label;
+          textEditingController.selection =
+              TextSelection.collapsed(offset: textEditingController.text.length);
+          currentHighlight = widget.enableSearch ? i : -1;
+        } : null,
+        requestFocusOnHover: false,
+        child: Text(entry.label),
+      );
+      result.add(menuItemButton);
+    }
+
+    return result;
+  }
+
+  void handlePressed(MenuController controller) {
+    if (controller.isOpen) {
+      currentHighlight = -1;
+      controller.close();
+    } else {  // close to open
+      if (_textEditingController.text.isNotEmpty) {
+        _enableFilter = false;
+      }
+      controller.open();
+    }
+    setState(() {});
+  }
+
+  @override
+  void dispose() {
+    _textEditingController.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final TextDirection textDirection = Directionality.of(context);
+    _initialMenu ??= _buildButtons(widget.comboBoxEntries, _textEditingController, textDirection);
+    final ComboBoxThemeData theme = ComboBoxTheme.of(context);
+    final ComboBoxThemeData defaults = _ComboBoxDefaultsM3(context);
+
+    if (_enableFilter) {
+      filteredEntries = filter(widget.comboBoxEntries, _textEditingController);
+    }
+
+    if (widget.enableSearch) {
+      currentHighlight = search(filteredEntries, _textEditingController);
+    }
+
+    final List<Widget> menu = _buildButtons(filteredEntries, _textEditingController, textDirection, focusedIndex: currentHighlight);
+
+    final TextStyle? effectiveTextStyle = widget.textStyle ?? theme.textStyle ?? defaults.textStyle;
+
+    MenuStyle? effectiveMenuStyle = widget.menuStyle
+      ?? theme.menuStyle
+      ?? defaults.menuStyle!;
+
+    final double? anchorWidth = getWidth(_anchorKey);
+    if (widget.width != null) {
+      effectiveMenuStyle = effectiveMenuStyle.copyWith(minimumSize: MaterialStatePropertyAll<Size?>(Size(widget.width!, 0.0)));
+    } else if (anchorWidth != null){
+      effectiveMenuStyle = effectiveMenuStyle.copyWith(minimumSize: MaterialStatePropertyAll<Size?>(Size(anchorWidth, 0.0)));
+    }
+
+    if (widget.menuHeight != null) {
+      effectiveMenuStyle = effectiveMenuStyle.copyWith(maximumSize: MaterialStatePropertyAll<Size>(Size(double.infinity, widget.menuHeight!)));
+    }
+    final InputDecorationTheme effectiveInputDecorationTheme = widget.inputDecorationTheme
+      ?? theme.inputDecorationTheme
+      ?? defaults.inputDecorationTheme!;
+
+    return Shortcuts(
+      shortcuts: _kMenuTraversalShortcuts,
+      child: Actions(
+        actions: <Type, Action<Intent>>{
+          _ArrowUpIntent: CallbackAction<_ArrowUpIntent>(
```

Comment:
These are complicated enough that it might be good to move them into their own private class or method.

Reply:
Created two methods `handleUpKeyInvoke` and `handleDownKeyInvoke` for these.